### PR TITLE
Add notes around usage with typescript

### PR DESCRIPTION
**Pull request by lorefnon**

### CODE REVIEW

**Comment by kibertoad** [comment]
@lorefnon Btw, considering that you already have contributor permissions for documentation repo, you don't need to use personal fork anymore, you can just branch out from gh-pages on main repo.

**Comment by lorefnon** [comment]
Yes, I'll do that for future PRs.

File: components/Code.jsx [hunk 1]
```diff
@@ -14,6 +14,7 @@ export default class Code extends Component {
     let fn
     if (language === 'js') fn = js
     if (language === 'sql') fn = sql
+    if (language === 'ts') fn = ts;
```
[the remark this answers is not in the document]
Comment (lorefnon) [reply components/Code.jsx]: I'll remove the semicolons manually for now.

File: sections/builder.js [hunk 2]
```diff
@@ -41,7 +41,83 @@ export default [
...
+          .where('id') // Your IDE will be able to help with the completion of id 
+          .first(); // Resolves to User | undefined
+      `
+    }, {
+      type: "heading",
+      size: "md",
+      content: "Caveat with type inferrence and mutable fluent APIs"
+    }, {
+      type: "text",
+      content: "Most of the knex APIs mutate current object and return it. This pattern does not work well with type-inferrence."
```
Comment (kibertoad) [inline sections/builder.js]: "inference"?

File: sections/builder.js [hunk 3]
```diff
@@ -80,6 +156,34 @@ export default [
...
         `
+      },
+      {
+        type: "heading",
+        size: "md",
+        content: "Usage with TypeScript"
+      },
+      {
+        type: "text",
+        content: "We are generally able to infer the result type based on the columns being selected as long as the select arguments match exactly the key names in record type. However aliasing and scoping can get in the way of inference."
```
Comment (kibertoad) [inline sections/builder.js]: However requires comma: https://onlinewritingtraining.com.au/2018/07/25/however-therefore-furthermore/

File: sections/typescript-support.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+export default [
+  {
+    type: "heading",
+    size: "lg",
+    content: "TypeScript Support",
+    href: "typescript-support"
+  },
+  {
+    type: "text",
+    content: "While knex is written in javascript, officially supported TypeScript bindings are available (within the knex npm package).",
```
Comment (kibertoad) [inline sections/typescript-support.js]: JavaScript for consistency?

File: sections/typescript-support.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  },
+  {
+    type: "text",
+    content: "However it is to be noted that TypeScript support is currently best-effort. Knex has a very flexible API and not all usage patterns can be type-checked and in most such cases we err on the side of flexibility. In particular, lack of type errors doesn't currently guarantee that the generated queries will be correct and therefore writing tests for them is recommended even if you are using TypeScript."
+  },
+  {
+    type: "text",
+    content: "Many of the APIs accept `TRecord` and `TResult` type parameters, using which we can specify the type of a row in the database table and the type of the result of the query respectively. This is helpful for auto-completion when using typescript aware editors like VSCode."
```
Comment (kibertoad) [inline sections/typescript-support.js]: TypeScript-aware?

**Comment by lorefnon** [comment]
I have taken care of spelling/grammar issues. The only other addition is incorporation of this [jsdoc example](https://github.com/tgriesser/knex/pull/3168#issuecomment-488454591) provided by @pvider.

**Comment by kibertoad** [comment]
Awesome, feel free to merge it!

**Pull request merged**